packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts

```diff
@@ -40,8 +64,109 @@ export const useGridInfiniteLoader = (apiRef: GridApiRef): void => {
       apiRef.current.publishEvent(GRID_ROWS_SCROLL_END, rowScrollEndParam);
       isInScrollBottomArea.current = true;
     }
-  }, [options, containerSizes, apiRef, visibleColumns]);
+  }, [logger, options, containerSizes, apiRef, visibleColumns]);
+
+  const handleGridVirtualPageChange = React.useCallback(
+    (params: GridVirtualPageChangeParams) => {
+      logger.debug('Virtual page changed');
+      if (!containerSizes || !options.getRows) {
+        return;
+      }
+
+      let nextPage = params.nextPage;
+
+      if (
+        params.nextPage > params.currentPage &&
+        params.nextPage !== renderState.renderedSizes!.lastPage
+      ) {
+        nextPage = params.nextPage + 1;
+      }
+
+      const newRowsBatchStartIndex = nextPage * containerSizes.viewportPageSize;
+      const toBeLoadedRange: any = [...allRows].splice(
+        newRowsBatchStartIndex,
+        containerSizes.viewportPageSize,
+      );
+
+      if (!toBeLoadedRange.includes(null)) {
+        return;
+      }
+
+      const { rows }: GridGetRowsReturnValue = options.getRows({
+        startIndex: newRowsBatchStartIndex,
+        viewportPageSize: containerSizes.viewportPageSize,
+        sortModel,
+        filterModel: filterState,
+        api: apiRef,
+      });
+
+      if (rows.length) {
+        apiRef.current.loadRows(newRowsBatchStartIndex, containerSizes.viewportPageSize, rows);
```

Comment:
There should two more rows before the loaded ones.

<img width="80" src="https://user-images.githubusercontent.com/42154031/119882641-03893f00-bf05-11eb-89e3-e3a7dc3ce134.png" alt="image" style="max-width:100%;">

Reply:
Updated